lightning/src/ln/channelmanager.rs

```diff
@@ -2120,164 +2246,152 @@ impl<Signer: Sign, M: Deref, T: Deref, K: Deref, F: Deref, L: Deref> ChannelMana
+					incoming_shared_secret: shared_secret,
+					amt_to_forward: $hop_data.amt_to_forward,
+					outgoing_cltv_value: $hop_data.outgoing_cltv_value,
+				})
+			}}
+		}
 
-			let blinding_factor = {
-				let mut sha = Sha256::engine();
-				sha.input(&new_pubkey.serialize()[..]);
-				sha.input(&shared_secret);
-				Sha256::from_engine(sha).into_inner()
-			};
+		let pending_forward_info = match next_hop {
+			onion_utils::Hop::Receive(next_hop_data) => {
+				// final_incorrect_cltv_expiry
+				if next_hop_data.outgoing_cltv_value != msg.cltv_expiry {
+					return_err!("Upstream node set CLTV to the wrong value", 18, &byte_utils::be32_to_array(msg.cltv_expiry));
+				}
+				get_recv_pending_htlc_status!(next_hop_data)
+			},
+			onion_utils::Hop::Forward { next_hop_data, next_hop_hmac, new_packet_bytes } => {
+				let mut new_pubkey = msg.onion_routing_packet.public_key.unwrap();
+
+				let blinding_factor = {
+					let mut sha = Sha256::engine();
+					sha.input(&new_pubkey.serialize()[..]);
+					sha.input(&shared_secret);
+					Sha256::from_engine(sha).into_inner()
+				};
 
-			let public_key = if let Err(e) = new_pubkey.mul_assign(&self.secp_ctx, &blinding_factor[..]) {
-				Err(e)
-			} else { Ok(new_pubkey) };
+				let public_key = if let Err(e) = new_pubkey.mul_assign(&self.secp_ctx, &blinding_factor[..]) {
+					Err(e)
+				} else { Ok(new_pubkey) };
 
-			let outgoing_packet = msgs::OnionPacket {
-				version: 0,
-				public_key,
-				hop_data: new_packet_data,
-				hmac: next_hop_hmac.clone(),
-			};
+				let outgoing_packet = msgs::OnionPacket {
+					version: 0,
+					public_key,
+					hop_data: new_packet_bytes,
+					hmac: next_hop_hmac.clone(),
+				};
 
-			let short_channel_id = match next_hop_data.format {
-				msgs::OnionHopDataFormat::Legacy { short_channel_id } => short_channel_id,
-				msgs::OnionHopDataFormat::NonFinalNode { short_channel_id } => short_channel_id,
-				msgs::OnionHopDataFormat::FinalNode { .. } => {
-					return_err!("Final Node OnionHopData provided for us as an intermediary node", 0x4000 | 22, &[0;0]);
-				},
-			};
+				let short_channel_id = match next_hop_data.format {
+					msgs::OnionHopDataFormat::Legacy { short_channel_id } => short_channel_id,
+					msgs::OnionHopDataFormat::NonFinalNode { short_channel_id } => short_channel_id,
+					msgs::OnionHopDataFormat::FinalNode { .. } => {
+						return_err!("Final Node OnionHopData provided for us as an intermediary node", 0x4000 | 22, &[0;0]);
+					},
+				};
 
-			PendingHTLCStatus::Forward(PendingHTLCInfo {
-				routing: PendingHTLCRouting::Forward {
-					onion_packet: outgoing_packet,
-					short_channel_id,
-				},
-				payment_hash: msg.payment_hash.clone(),
-				incoming_shared_secret: shared_secret,
-				amt_to_forward: next_hop_data.amt_to_forward,
-				outgoing_cltv_value: next_hop_data.outgoing_cltv_value,
-			})
+				let phantom_secret_res = self.keys_manager.get_phantom_secret(short_channel_id);
```

Comment:
Hmmmmmmm, I do wonder if we should care about "any third party can identify if a hop is a phantom node or not". If its too hard to do, we can probably call it a wash and not bother, but I'd ideally like to at least try to obfuscate it first. I think it should be pretty reasonable here - instead of decoding the second onion here immediately we instead push the HTLC through `PendingHTLCStatus::Forward` first, which means we wait until we've gotten it fully committed in the channel and then do this second-step decode in `process_pending_htlc_forwards` where we look up the SCID in the loop.

Reply:
Added passing it through `process_pending_htlc_forwards` rather than decoding the second onion immediately.

> "any third party can identify if a hop is a phantom node or not"

To be clear, this is a timing attack?

Reply:
> To be clear, this is a timing attack?

Yes. Hopefully pushing it through the normal forward-batching, in theory, implies timing is similar across forward-one-hop's and no-forward, but its probably still wide enough just with the skipping of network RTTs and disk updates that we should tell users they can expect a counterparty to detect the difference.